[mlir][tensor] Add a tensor.concat operation

qedawkins opened the pull request:
This adds an operation for concatenating ranked tensors along a static dimension, as well as a decomposition mirroring the existing lowering from TOSA to Tensor. This offers a convergence point for "input" like dialects that include various lowerings for concatenation operations, easing later analysis. In the future, this op can implement the necessary interfaces for tiling, as well as potentially add conversions to some kind of linalg and/or memref counterpart.

This patch adds the op, the decomposition, and some basic folding/canonicalization. Replacing lowerings with the op (such as the TOSA lowering) will come as a follow up.

See https://discourse.llvm.org/t/rfc-tensor-add-a-tensor-concatenate-operation/74858

qedawkins committed:
[mlir][tensor] Add tensor.concat operation

This adds an operation for concatenating ranked tensors along a static
dimension, as well as a decomposition mirroring the existing lowering
from Tosa to Tensor. This offers for "input" like dialects that include
various lowerings for concatenation operations, easing later analysis.
In the future, this op can implement the necessary interfaces for
tiling, as well as potentially add conversions to some kind of linalg
and/or memref counterpart.

See https://discourse.llvm.org/t/rfc-tensor-add-a-tensor-concatenate-operation/74858

qedawkins committed:
Fix reifyResultShapes implementation

nicolasvasilache commented on mlir/lib/Dialect/Tensor/IR/TensorOps.cpp:
```diff
@@ -471,6 +471,202 @@ void CastOp::getCanonicalizationPatterns(RewritePatternSet &results,
+FailureOr<RankedTensorType> ConcatOp::inferResultType(int64_t dim,
+                                                      TypeRange inputTypes) {
+  if (dim < 0)
+    return failure();
```
It seems un some of these cases, you're trying a bit too hard to be nice and end up duplicating the verifier.
Maybe we can turn some of these failures into assertions (but being very careful that they cannot be hit from verifier because that is another can of worms..)

hanhanW replied:
+1. I think we can move the checks to verifier and make this method return `RankedTensorType` type. Like other static methods, user should be careful when they use the util. It is okay to not return a failure in this static method. Adding assertions is good enough to me.

nicolasvasilache commented on mlir/lib/Dialect/Tensor/IR/TensorOps.cpp:
```diff
@@ -471,6 +471,202 @@ void CastOp::getCanonicalizationPatterns(RewritePatternSet &results,
+
+  SmallVector<int64_t> sizes(init.getShape());
+  Type elementType = init.getElementType();
+  for (Type type : inputTypes.drop_front()) {
```
If it were me, I would probably structure this differently:
1. for all `ranks != dim`, if `llvm::any_of tensor.shape()[rank]` mismatches return failure with a proper emitOpError msg.
2. for `dim`, expose/evolve and reuse "saturated_arith" helpers from here: https://github.com/llvm/llvm-project/blob/cfee7152d4eb673976b51b831295dcf5b1811634/mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp#L32

It seems people routinely rewrite logic around isDynamic etc that we would be better off commonalizing.

qedawkins replied:
I moved it and reused it. Let me know if that's what you were looking for (and if the naming makes sense).

nicolasvasilache commented on mlir/lib/Dialect/Tensor/IR/TensorOps.cpp:
```diff
@@ -471,6 +471,202 @@ void CastOp::getCanonicalizationPatterns(RewritePatternSet &results,
+  // from the given result type, as well as the inferred result type, otherwise
+  // use the dim sizes from the first input.
+  bool hasStaticConcatDim = false;
+  for (int64_t i = 0; i < rank; ++i) {
```
I would prob. also structure this differently to make it simpler to read: 
1. for all `ranks != dim`, copy the shape of the first tensor (assuming the interplay with the verifier is good)
2. for `dim`, take the static result size of the sum.

Looking deeper, it seems similar to what you are implementing here, I'd just expect this to take ~10 lines and be very easy to parse to human eyes (vs the ~40 I see here).

nicolasvasilache commented on mlir/lib/Dialect/Tensor/Transforms/ConcatOpPatterns.cpp:
```diff
@@ -0,0 +1,86 @@
+    for (auto input : concatOp.getInputs().drop_front()) {
+      Value size = rewriter.createOrFold<tensor::DimOp>(loc, input, dimValue);
+      Value currentOffset =
+          getValueOrCreateConstantIndexOp(rewriter, loc, dimOffsets.back());
```
We should already have better helpers that use `OpFoldResult` and `affine_apply` ops to make this more idiomatic and more concise IR.

hanhanW replied:
+1, can we use affine ops?

nicolasvasilache commented on mlir/test/lib/Dialect/Tensor/TestTensorTransforms.cpp:
```diff
@@ -77,6 +77,10 @@ struct TestTensorTransforms
       llvm::cl::desc("Test folding ops into tensor.pack and tensor.unpack"),
       llvm::cl::init(false)};
 
+  Option<bool> testDecomposeConcat{
```
Can we plug this into a simple TD ApplyPattern rather than add more to this one-off C++ test ?
It is both shorther and more localized.

qedawkins replied:
Sounds good to me! I just wasn't sure if we were happy with having testing stories being tied to transform ops.

nicolasvasilache replied:
In general, I think we should move to transform ops for everything that requires such custom C++ test logic and make sense as a transform op.
C++ tests that are modeled like this one:
  1. are throwaway and never used outside of test, at least transforms are reusable.
  2. are opaque and hidden behind a name, at least transforms make the action of the test very clear.
  3. have intertwined `runOnOperation` implementation that just keeps on growing

nicolasvasilache reviewed (approved):
Thanks!

qedawkins commented on mlir/include/mlir/Dialect/Tensor/Transforms/Transforms.h:
```diff
@@ -67,6 +67,11 @@ void populateReassociativeReshapeFoldingPatterns(RewritePatternSet &patterns);
+/// Populates `patterns` with patterns that decompose `tensor.concat` into
+/// `tensor.empty` of a tensor of the concatenated size, followed by a chain
+/// of `tensor.insert_slice` operations on the inputs.
+void populateDecomposeTensorConcatPatterns(RewritePatternSet &patterns);
```
Sure. Should I split out the implementation of `ReifyRankedShapedTypeOpInterface` as well? I'm not sure how to test this interface in isolation.

qedawkins commented on mlir/include/mlir/Dialect/Tensor/IR/TensorOps.td:
````diff
@@ -121,6 +121,66 @@ def Tensor_CastOp : Tensor_Op<"cast", [
+    Example:
+
+    ```mlir
+    %0 = tensor.concat dim(0) %0, %1, %2 :
````
I was expecting the same as `tensor.pad` where there are ways to rewrite it as an equivalent set of ops in DPS (either as a single linalg op or as a group of linalg ops). Bufferization would either require the DPS rewrite in advance (although that linalg op is quite ugly), or we allocate based on the result shape and copy one-by-one into the allocated buffer.

I'm not married to leaving it as non-DPS. It's just more consistent with other input dialects and might make higher level transformations easier.

qedawkins replied:
After some more thought, if we ever wanted a `memref.concat`, I would much prefer having a destination on `tensor.concat` to get there. If others agree, I'm open to switching it to DPS and just providing nice helpers for constructing the destination (so that we don't duplicate that logic in every frontend lowering).

MaheshRavishankar replied:
I dont think `memref.concat` makes sense. The norm is that `memref` operations dont have copy semantics, so `memref.concat` might not be expressible as that (without introducing some pointer arithmetic stuff). I think at `memref` level `memref.subviews` + `memref.copy` are better representation of concat, with something else converting the `tensor.concat` into as much as in-place update as possible. Given that I think not having DPS for `concat` might be fine...

qedawkins replied:
I was assuming a `memref.concat` would take the destination and implicitly just be that sequence of subviews + copies. Maybe that is just a convenience op and not really worth adding, or `linalg.concat` would be preferred. The reason I'm thinking about wanting DPS now is if we wanted to tile then bufferize. Without the destination I would expect concat to turn into an alloc + sequence of copies. With a destination, we're concatenating into a slice.

MaheshRavishankar replied:
> I was assuming a `memref.concat` would take the destination and implicitly just be that sequence of subviews + copies. Maybe that is just a convenience op and not really worth adding, or `linalg.concat` would be preferred. The reason I'm thinking about wanting DPS now is if we wanted to tile then bufferize. Without the destination I would expect concat to turn into an alloc + sequence of copies. With a destination, we're concatenating into a slice.

ok, that is a fair point. We have two examples of both. One is `tensor.pad` that is not destination passing style and you have to do something special to turn it into destination passing style. We could do that, or we do what `tensor.pack` does which is destination passing style. On balance I would maybe go with DPS for what you suggest.

matthias-springer replied:
No strong opinion from my side regarding DPS/non-DPS. For the sake of simplicity (and to some degree consistency with the tensor dialect; most ops in there are non-DPS), I'd slightly prefer a non-DPS variant that allocates when bufferizing it. We can lower the op to `tensor.empty` + multiple `tensor.insert_slice`, then use empty tensor elimination to infer the destination. Basically the same way we handle `tensor.pad` today.

I've been thinking of extending `bufferization.materialize_in_destination` with a region that can wrap arbitrary non-DPS op and provide the missing destination. That could be an alternative to turning more and more ops into DPS style.

Regarding `memref.concat`: I would just generate `memref.subview`+`memref.copy` directly to keep the memref dialect as simple as possible. (Unless you want to do tiling of memref IR.)

joker-eph commented on mlir/include/mlir/Dialect/Tensor/Transforms/Transforms.h:
```diff
@@ -67,6 +67,11 @@ void populateReassociativeReshapeFoldingPatterns(RewritePatternSet &patterns);
+/// Populates `patterns` with patterns that decompose `tensor.concat` into
+/// `tensor.empty` of a tensor of the concatenated size, followed by a chain
+/// of `tensor.insert_slice` operations on the inputs.
+void populateDecomposeTensorConcatPatterns(RewritePatternSet &patterns);
```
The document says "what" but I feel could expand on the "why".

Most of these documentations for these APIs aren't useful enough because they don't provide enough context I feel.

hanhanW commented on mlir/lib/Dialect/Tensor/IR/TensorOps.cpp:
```diff
@@ -471,6 +471,202 @@ void CastOp::getCanonicalizationPatterns(RewritePatternSet &results,
+  SmallVector<int64_t> sizes(init.getShape());
+  Type elementType = init.getElementType();
+  for (Type type : inputTypes.drop_front()) {
+    RankedTensorType tensorType = dyn_cast<RankedTensorType>(type);
```
style nit: use auto because dyn_cast already spells the type.

hanhanW commented on mlir/lib/Dialect/Tensor/IR/TensorOps.cpp:
```diff
@@ -471,6 +471,202 @@ void CastOp::getCanonicalizationPatterns(RewritePatternSet &results,
+
+  RankedTensorType resultType = getResultType();
+
+  int64_t resultRank = resultType.getRank();
```
[optional] I think adding a `getRank()` method to concate op is useful.

hanhanW commented on mlir/lib/Dialect/Tensor/Transforms/ConcatOpPatterns.cpp:
```diff
@@ -0,0 +1,86 @@
+
+    Location loc = concatOp.getLoc();
+    FailureOr<Value> dest =
+        tensor::getOrCreateDestination(rewriter, loc, concatOp->getResult(0));
```
Can we use the ConcatOp result getter instead of the one from `Operation *`? I.e., does `concatOp.getResult()` work?

qedawkins replied:
This internally calls `ReifyResultShapes` to get the shape of the concatenated tensor, and then constructs the appropriate `tensor.empty`. Currently `tensor.concat` is non-DPS so we need to invoke that logic to construct the result. We could add a wrapper around this as a class method though.

hanhanW commented on mlir/lib/Dialect/Tensor/Transforms/ConcatOpPatterns.cpp:
```diff
@@ -0,0 +1,86 @@
+
+    Value result = *dest;
+
+    for (auto [input, offset] : llvm::zip(concatOp.getInputs(), dimOffsets)) {
```
nit: I prefer using `zip_equal` when possible. It is self-documenting and stricter.

hanhanW commented on mlir/test/Dialect/Tensor/ops.mlir:
```diff
@@ -15,6 +15,23 @@ func.func @cast(%arg0: tensor<*xf32>, %arg1 : tensor<4x4xf32>, %arg2: tensor<?x?
+  // CHECK: tensor.concat dim(1) %{{.*}} : (tensor<?x?x?xf32>, tensor<?x?x?xf32>) -> tensor<?x10x?xf32>
+  %3 = tensor.concat dim(1) %arg2, %arg2 : (tensor<?x?x?xf32>, tensor<?x?x?xf32>) -> tensor<?x10x?xf32>
+  // CHECK: tensor.concat dim(1) %{{.*}} : (tensor<?x?x?xf32>, tensor<4x4x3xf32>, tensor<4x7x3xf32>) -> tensor<4x?x3xf32>
+  %4 = tensor.concat dim(1) %arg2, %arg1, %arg0 : (tensor<?x?x?xf32>, tensor<4x4x3xf32>, tensor<4x7x3xf32>) -> tensor<4x?x3xf32>
```
[optional] The test looks interesting because we are able to infer some shapes! Perhaps we can have a follow-up that inserts some tensor.cast ops around for helping shape inference. We have [similar features](https://github.com/llvm/llvm-project/blob/fea023b129190edeb503dbe947034f925bbda666/mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp#L2264) in Linalg dialect. Do you consider to add such inference to tensor.concat canonicalization patterns?

qedawkins committed:
Rebase and address comments

- Simplify various implementations
- Make inferResultType always return RankedTensorType
- Import saturated_arith and reuse for concat
- Clarify reason for decomposing
- Use Affine folding helpers
- Switch test to a transform op

hanhanW commented on mlir/lib/Dialect/Tensor/IR/TensorOps.cpp:
```diff
@@ -471,6 +471,192 @@ void CastOp::getCanonicalizationPatterns(RewritePatternSet &results,
+  // The concatenation dim must be in the range [0, rank).
+  assert(dim >= 0 && dim < concatRank && "Invalid concatenation dim");
+
+  SmallVector<int64_t> sizes((concatRank));
```
nit: we can wrap it with a single `()` pair.

```suggestion
  SmallVector<int64_t> sizes(concatRank);
```

hanhanW commented on mlir/lib/Dialect/Tensor/IR/TensorOps.cpp:
```diff
@@ -471,6 +471,192 @@ void CastOp::getCanonicalizationPatterns(RewritePatternSet &results,
+  if (dim >= resultRank)
+    return emitOpError("concatenation dim must be less than the tensor rank");
+
+  SmallVector<int64_t> sizes((resultRank));
```
ditto

```suggestion
  SmallVector<int64_t> sizes(resultRank);
```

hanhanW commented on mlir/lib/Dialect/Tensor/IR/TensorOps.cpp:
```diff
@@ -471,6 +471,192 @@ void CastOp::getCanonicalizationPatterns(RewritePatternSet &results,
+
+  // Take the sum of the input sizes along the concatenated dim.
+  AffineExpr sum = builder.getAffineDimExpr(0);
+  SmallVector<OpFoldResult> sizes{
+      builder.create<tensor::DimOp>(init.getLoc(), init, 0).getResult()};
```
Can we use assignment syntax? The brace initialization with constructor syntax is dangerous. It could potentially lead to a bug, e.g.,

```
std::vector<std::string> strings{2}; // A vector of two empty strings.
std::vector<int> ints{2};            // A vector containing only the integer 2.
```

For more context, see https://abseil.io/tips/88

hanhanW commented on mlir/lib/Dialect/Tensor/Transforms/ConcatOpPatterns.cpp:
```diff
@@ -0,0 +1,93 @@
+
+    // Compute the partial sums for the slice offsets.
+    AffineExpr sum = rewriter.getAffineDimExpr(0);
+    SmallVector<AffineExpr> partialSums{sum};
+    SmallVector<OpFoldResult> offsetStrides{rewriter.getIndexAttr(0)};
```
ditto, can we use assignment syntax?

```suggestion
    SmallVector<AffineExpr> partialSums = {sum};
    SmallVector<OpFoldResult> offsetStrides = {rewriter.getIndexAttr(0)};
```

hanhanW reviewed (approved):
Thanks for pushing on it, just few style nits!

qedawkins committed:
Address comments

qedawkins committed:
buildbot can't find my commit :(
